Build issue on Solus with glibc debug symbols

I am getting this error while trying to run the build script.
 
```
jrud@solus_thinkpad ~/documents/school/thesis/shadow $ ./setup build
2017-10-07 13:11:58,469 [INFO] running 'cmake /home/jrud/documents/school/thesis/shadow -DCMAKE_INSTALL_PREFIX=/home/jrud/.shadow -DCMAKE_EXTRA_INCLUDES=/home/jrud/documents/school/thesis/shadow/build;/home/jrud/.shadow/include -DCMAKE_EXTRA_LIBRARIES=/home/jrud/documents/school/thesis/shadow/build;/home/jrud/.shadow/lib' from '/home/jrud/documents/school/thesis/shadow/build'
-- The C compiler identification is Clang 4.0.1
-- Check for working C compiler: /usr/bin/clang
-- Check for working C compiler: /usr/bin/clang -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- 
-- -------------------------------------------------------------------------------
-- Current settings: (change with '$ cmake -D<OPTION>=<ON|OFF>')
-- SHADOW_DEBUG=OFF
-- SHADOW_PROFILE=OFF
-- SHADOW_TEST=OFF
-- SHADOW_EXPORT=OFF
-- -------------------------------------------------------------------------------
-- 
-- CMAKE_BUILD_TYPE Release enabled.
-- Check if the system is big endian
-- Searching 16 bit integer
-- Looking for sys/types.h
-- Looking for sys/types.h - found
-- Looking for stdint.h
-- Looking for stdint.h - found
-- Looking for stddef.h
-- Looking for stddef.h - found
-- Check size of unsigned short
-- Check size of unsigned short - done
-- Using unsigned short
-- Check if the system is big endian - little endian
-- CMAKE_MODULE_PATH = /home/jrud/documents/school/thesis/shadow/cmake/;/home/jrud/.shadow/include;/home/jrud/documents/school/thesis/shadow/build
-- Decompressing /home/jrud/documents/school/thesis/shadow/resource/topology.graphml.xml.xz
-- Found components for RT
-- RT_INCLUDES = /usr/include
-- RT_LIBRARIES = /usr/lib/librt.so
-- Found components for DL
-- DL_INCLUDES = /usr/include
-- DL_LIBRARIES = /usr/lib/libdl.so
-- Found components for M
-- M_INCLUDES = /usr/include
-- M_LIBRARIES = /usr/lib/libm.so
-- Found components for IGRAPH
-- IGRAPH_INCLUDES = /usr/local/include/igraph
-- IGRAPH_LIBRARIES = /usr/local/lib/libigraph.so
-- IGRAPH_VERSION_STRING_LINE = #define IGRAPH_VERSION "0.7.1"
-- IGRAPH_VERSION_MAJOR_GUESS = 0
-- IGRAPH_VERSION_MINOR_GUESS = 7
-- IGRAPH_VERSION_PATCH_GUESS = 1
-- Found components for GLIB
-- GLIB_INCLUDES = /usr/include/glib-2.0;/usr/include/glib-2.0
-- GLIB_LIBRARIES = /usr/lib/libglib-2.0.so;/usr/lib/libgthread-2.0.so;/usr/lib/libgmodule-2.0.so
-- GLIB_MAJOR_VERSION = 
-- GLIB_MINOR_VERSION = 
-- GLIB_MICRO_VERSION = 
-- Checking for C LLVM compiler...
-- Checking for C LLVM compiler... works.
-- Checking for CXX LLVM compiler...
-- Checking for CXX LLVM compiler... works.
-- The CXX compiler identification is Clang 4.0.1
-- Check for working CXX compiler: /usr/bin/clang++
-- Check for working CXX compiler: /usr/bin/clang++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found components for LLVM
-- LLVM_CMAKE_DIR = /usr/lib/cmake/llvm
CMake Warning (dev) at /usr/lib/cmake/llvm/AddLLVM.cmake:629 (if):
  Policy CMP0057 is not set: Support new IN_LIST if() operator.  Run "cmake
  --help-policy CMP0057" for policy details.  Use the cmake_policy command to
  set the policy and suppress this warning.

  IN_LIST will be interpreted as an operator when the policy is set to NEW.
  Since the policy is not set the OLD behavior will be used.
Call Stack (most recent call first):
  src/plugin/llvm-hoist-pass/CMakeLists.txt:44 (add_llvm_loadable_module)
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Error at /usr/lib/cmake/llvm/AddLLVM.cmake:629 (if):
  if given arguments:

    "LLVMHoistGlobals" "IN_LIST" "LLVM_DISTRIBUTION_COMPONENTS" "OR" "NOT" "LLVM_DISTRIBUTION_COMPONENTS"

  Unknown arguments specified
Call Stack (most recent call first):
  src/plugin/llvm-hoist-pass/CMakeLists.txt:44 (add_llvm_loadable_module)


-- Configuring incomplete, errors occurred!
See also "/home/jrud/documents/school/thesis/shadow/build/CMakeFiles/CMakeOutput.log".
2017-10-07 13:12:01,380 [ERROR]  Non-zero return code from cmake.
```
CMake version:
```
jrud@solus_thinkpad ~/documents/school/thesis/shadow $ cmake --version
cmake version 3.8.2

CMake suite maintained and supported by Kitware (kitware.com/cmake).
```
Clang version:
```
jrud@solus_thinkpad ~/documents/school/thesis/shadow $ clang --version
clang version 4.0.1 (tags/RELEASE_401/final)
Target: x86_64-solus-linux
Thread model: posix
InstalledDir: /usr/bin
```
Distro info: 
```
jrud@solus_thinkpad ~/documents/school/thesis/shadow $ lsb_release -a
LSB Version:	1.4
Distributor ID:	Solus
Description:	Solus
Release:	3
Codename:	shannon
```
Linux version:
```
jrud@solus_thinkpad ~/documents/school/thesis/shadow $ uname -srv
Linux 4.12.14-21.current #1 SMP Fri Sep 22 03:15:23 UTC 2017
```

OK so I pulled down a clean version of master, and this is the result now:

```
jrud@solus_thinkpad ~/documents/school/thesis/shadow $ ./setup build
2017-10-09 12:51:32,530 [INFO] running 'cmake /home/jrud/documents/school/thesis/shadow -DCMAKE_INSTALL_PREFIX=/home/jrud/.shadow -DCMAKE_EXTRA_INCLUDES=/home/jrud/documents/school/thesis/shadow/build;/home/jrud/.shadow/include -DCMAKE_EXTRA_LIBRARIES=/home/jrud/documents/school/thesis/shadow/build;/home/jrud/.shadow/lib' from '/home/jrud/documents/school/thesis/shadow/build'
-- The C compiler identification is GNU 6.4.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- System name: Linux
-- System version: 4.12.14-21.current
-- System processor: x86_64
-- CMAKE_MODULE_PATH = /home/jrud/documents/school/thesis/shadow/cmake/;/home/jrud/.shadow/include;/home/jrud/documents/school/thesis/shadow/build
-- Check if the system is big endian
-- Searching 16 bit integer
-- Looking for sys/types.h
-- Looking for sys/types.h - found
-- Looking for stdint.h
-- Looking for stdint.h - found
-- Looking for stddef.h
-- Looking for stddef.h - found
-- Check size of unsigned short
-- Check size of unsigned short - done
-- Using unsigned short
-- Check if the system is big endian - little endian
-- 
-- -------------------------------------------------------------------------------
-- Current settings: (change with '$ cmake -D<OPTION>=<ON|OFF>')
-- SHADOW_DEBUG=OFF
-- SHADOW_PROFILE=OFF
-- SHADOW_TEST=OFF
-- SHADOW_EXPORT=OFF
-- -------------------------------------------------------------------------------
-- 
-- CMAKE_BUILD_TYPE Release enabled.
-- Decompressing /home/jrud/documents/school/thesis/shadow/resource/topology.graphml.xml.xz
-- Found components for RT
-- RT_INCLUDES = /usr/include
-- RT_LIBRARIES = /usr/lib/librt.so
-- Found components for DL
-- DL_INCLUDES = /usr/include
-- DL_LIBRARIES = /usr/lib/libdl.so
-- Found components for M
-- M_INCLUDES = /usr/include
-- M_LIBRARIES = /usr/lib/libm.so
-- Found components for IGRAPH
-- IGRAPH_INCLUDES = /usr/local/include/igraph
-- IGRAPH_LIBRARIES = /usr/local/lib/libigraph.so
-- IGRAPH_VERSION_STRING_LINE = #define IGRAPH_VERSION "0.7.1"
-- IGRAPH_VERSION_MAJOR_GUESS = 0
-- IGRAPH_VERSION_MINOR_GUESS = 7
-- IGRAPH_VERSION_PATCH_GUESS = 1
-- Found components for GLIB
-- GLIB_INCLUDES = /usr/include/glib-2.0;/usr/include/glib-2.0
-- GLIB_LIBRARIES = /usr/lib/libglib-2.0.so;/usr/lib/libgthread-2.0.so;/usr/lib/libgmodule-2.0.so
-- GLIB_MAJOR_VERSION = 
-- GLIB_MINOR_VERSION = 
-- GLIB_MICRO_VERSION = 
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Using Threads library: -lpthread
-- Found components for GLIB
-- GLIB_INCLUDES = /usr/include/glib-2.0;/usr/include/glib-2.0
-- GLIB_LIBRARIES = /usr/lib/libglib-2.0.so;/usr/lib/libgthread-2.0.so;/usr/lib/libgmodule-2.0.so
-- GLIB_MAJOR_VERSION = 
-- GLIB_MINOR_VERSION = 
-- GLIB_MICRO_VERSION = 
-- Found components for IGRAPH
-- IGRAPH_INCLUDES = /usr/local/include/igraph
-- IGRAPH_LIBRARIES = /usr/local/lib/libigraph.so
-- IGRAPH_VERSION_STRING_LINE = #define IGRAPH_VERSION "0.7.1"
-- IGRAPH_VERSION_MAJOR_GUESS = 0
-- IGRAPH_VERSION_MINOR_GUESS = 7
-- IGRAPH_VERSION_PATCH_GUESS = 1
-- Found components for DL
-- DL_INCLUDES = /usr/include
-- DL_LIBRARIES = /usr/lib/libdl.so
-- The CXX compiler identification is GNU 6.4.0
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found components for RT
-- RT_INCLUDES = /usr/include
-- RT_LIBRARIES = /usr/lib/librt.so
-- Found components for M
-- M_INCLUDES = /usr/include
-- M_LIBRARIES = /usr/lib/libm.so
-- Found components for IGRAPH
-- IGRAPH_INCLUDES = /usr/local/include/igraph
-- IGRAPH_LIBRARIES = /usr/local/lib/libigraph.so
-- IGRAPH_VERSION_STRING_LINE = #define IGRAPH_VERSION "0.7.1"
-- IGRAPH_VERSION_MAJOR_GUESS = 0
-- IGRAPH_VERSION_MINOR_GUESS = 7
-- IGRAPH_VERSION_PATCH_GUESS = 1
-- Found components for GLIB
-- GLIB_INCLUDES = /usr/include/glib-2.0;/usr/include/glib-2.0
-- GLIB_LIBRARIES = /usr/lib/libglib-2.0.so;/usr/lib/libgthread-2.0.so;/usr/lib/libgmodule-2.0.so
-- GLIB_MAJOR_VERSION = 
-- GLIB_MINOR_VERSION = 
-- GLIB_MICRO_VERSION = 
-- CMAKE_C_FLAGS = -g2 -O3 -pipe -fPIC -Wformat -Wformat-security -fno-omit-frame-pointer -fexceptions -D_FORTIFY_SOURCE=2 -fstack-protector --param ssp-buffer-size=32 -fasynchronous-unwind-tables -ftree-vectorize -feliminate-unused-debug-types -Wall -Wno-error -Wp,-D_REENTRANT -O3 -Wno-unknown-attributes -Wno-unused-command-line-argument -Wno-unknown-warning-option -Wno-invalid-noreturn -fPIC -std=gnu11 -fno-inline -fno-strict-aliasing -U_FORTIFY_SOURCE
-- The ASM compiler identification is GNU
-- Found assembler: /usr/bin/cc
-- System version: 4.12.14-21.current
-- Architecture: x86_64
-- LDSO = /lib64/ld-linux-x86-64.so.2
-- LIBDL = 
-- Configuring done
-- Generating done
-- Build files have been written to: /home/jrud/documents/school/thesis/shadow/build
2017-10-09 12:51:35,642 [INFO] calling make -j1
Scanning dependencies of target vdlconfig
[  1%] Generating vdl-config.h
Traceback (most recent call last):
  File "/home/jrud/documents/school/thesis/shadow/src/external/elf-loader/extract-system-config.py", line 268, in <module>
    main(sys.argv[1:])
  File "/home/jrud/documents/school/thesis/shadow/src/external/elf-loader/extract-system-config.py", line 207, in main
    debug_filename = search_debug_file ()
  File "/home/jrud/documents/school/thesis/shadow/src/external/elf-loader/extract-system-config.py", line 166, in search_debug_file
    raise CouldNotFindFile ()
__main__.CouldNotFindFile: <__main__.CouldNotFindFile instance at 0x7f002f2d8a70>
make[2]: *** [src/external/elf-loader/CMakeFiles/vdlconfig.dir/build.make:61: src/external/elf-loader/vdl-config.h] Error 1
make[2]: *** Deleting file 'src/external/elf-loader/vdl-config.h'
make[1]: *** [CMakeFiles/Makefile2:711: src/external/elf-loader/CMakeFiles/vdlconfig.dir/all] Error 2
make: *** [Makefile:152: all] Error 2
2017-10-09 12:51:35,703 [ERROR] Non-zero return code from make.

```

I installed Solus Budgie on a VM and corrected the search for `ld-2.25.so` in ffe880f638e7e2ced04fcdbdde80f84c3693ab6d.

The next error is:

```
Scanning dependencies of target vdlconfig
[  2%] Generating vdl-config.h
Traceback (most recent call last):
  File "/home/rob/shadow/src/external/elf-loader/extract-system-config.py", line 295, in <module>
    main(sys.argv[1:])
  File "/home/rob/shadow/src/external/elf-loader/extract-system-config.py", line 256, in main
    data = debug.get_struct_member_offset ('rtld_global', '_dl_error_catch_tsd')
  File "/home/rob/shadow/src/external/elf-loader/extract-system-config.py", line 122, in get_struct_member_offset
    return self.find_member (member_name, item)
  File "/home/rob/shadow/src/external/elf-loader/extract-system-config.py", line 108, in find_member
    self.write_back_one ()
TypeError: write_back_one() takes exactly 2 arguments (1 given)
make[2]: *** [src/external/elf-loader/CMakeFiles/vdlconfig.dir/build.make:61: src/external/elf-loader/vdl-config.h] Error 1
make[2]: *** Deleting file 'src/external/elf-loader/vdl-config.h'
make[1]: *** [CMakeFiles/Makefile2:721: src/external/elf-loader/CMakeFiles/vdlconfig.dir/all] Error 2
make: *** [Makefile:163: all] Error 2
2017-10-18 10:29:48,749 [ERROR] Non-zero return code from make.
```

--- a/src/external/elf-loader/CMakeLists.txt
+++ b/src/external/elf-loader/CMakeLists.txt
@@ -38,6 +38,8 @@ if( ${CMAKE_SYSTEM_PROCESSOR} STREQUAL "x86_64" )
         set(LIBDL_FILE /lib/x86_64-linux-gnu/libdl.so.2)
     elseif(EXISTS /lib64/libdl.so.2)
         set(LIBDL_FILE /lib64/libdl.so.2)
+    elseif(EXISTS /usr/lib64/libdl.so.2)
+        set(LIBDL_FILE /usr/lib64/libdl.so.2)
     endif()
     message(STATUS "LIBDL = ${LIBDL_FILE}")
 

--- a/src/external/elf-loader/extract-system-config.py
+++ b/src/external/elf-loader/extract-system-config.py
@@ -254,9 +254,9 @@ def main(argv):
     config.write ('#define CONFIG_RTLD_DL_CLKTCK_OFFSET ' + str(data.data) + '\n')
 
     data = debug.get_struct_member_offset ('rtld_global', '_dl_error_catch_tsd')
-    if data is None:
-        sys.exit (1)
-    config.write ('#define CONFIG_DL_ERROR_CATCH_TSD_OFFSET ' + str(data.data) + '\n')
+    # field was removed in glibc 2.25
+    if data is not None:
+        config.write ('#define CONFIG_DL_ERROR_CATCH_TSD_OFFSET ' + str(data.data) + '\n')
 
     data = debug.get_struct_size ('pthread')
     if data is None:

--- a/src/external/elf-loader/extract-system-config.py
+++ b/src/external/elf-loader/extract-system-config.py
@@ -175,6 +175,9 @@ def search_debug_file():
                     # ubuntu 1610
                     '/usr/lib/debug/lib/x86_64-linux-gnu/ld-2.24.so',
                     '/usr/lib/debug/lib/i386-linux-gnu/ld-2.24.so',
+                    # ubuntu 1710
+                    '/usr/lib/debug/lib/x86_64-linux-gnu/ld-2.26.so',
+                    '/usr/lib/debug/lib/i386-linux-gnu/ld-2.26.so',
                     # debian 9
                     find_build_id('/lib/x86_64-linux-gnu/ld-2.24.so'),
                     find_build_id('/lib/i386-linux-gnu/ld-2.24.so'),

--- a/src/external/elf-loader/extract-system-config.py
+++ b/src/external/elf-loader/extract-system-config.py
@@ -105,7 +105,7 @@ class DebugData:
         sub_item = self.read_one ()
         while sub_item is not None:
             if sub_item.level == parent.level:
-                self.write_back_one ()
+                self.write_back_one (sub_item)
                 return None
             if sub_item.type.strip() == 'DW_TAG_member' and \
                     sub_item.attributes.has_key ('DW_AT_name') and \
@@ -187,12 +187,14 @@ def search_debug_file():
 
         if os.path.isfile (file):
             return file
-    
+
     raise CouldNotFindFile ()
 
 def list_lib_path():
     paths = []
     re_lib = re.compile ('(?<=^#)')
+    if not os.path.isdir("/etc/ld.so.conf.d/"):
+        return ''
     for filename in os.listdir("/etc/ld.so.conf.d/"):
         try:
             for line in open ("/etc/ld.so.conf.d/" + filename, 'r'):
@@ -202,13 +204,14 @@ def list_lib_path():
         except:
             continue
     return ':'.join(paths)
-        
+
 def usage():
     print ''
 
 def main(argv):
     config_filename = ''
     debug_filename = ''
+    build_dir = ''
     try:
         opts, args = getopt.getopt(argv, 'hc:d:b:',
                                    ['help', 'config=', 'debug=', 'builddir='])
@@ -293,4 +296,3 @@ def main(argv):
 
 if __name__ == "__main__":
     main(sys.argv[1:])
-

--- a/src/external/elf-loader/glibc.c
+++ b/src/external/elf-loader/glibc.c
@@ -80,6 +80,38 @@ EXPORT char **_dl_argv;
 //_r_debug;
 //__libc_memalign;
 
+// XXX: tunables were added in 2.25. They allow customization of glibc
+// behavior via environment variables in a more (though not completely)
+// standardized manner. A good place to start to understand them is here:
+// https://siddhesh.in/posts/the-story-of-tunables.html
+// Unfortunately, they are implemented in the dynamic loader, and while
+// they currently aren't heavily used, that's likely to change. For now,
+// we get away with some stub code that seems to get us far enough, but
+// what we really need to do is copy the entire implementation here.
+typedef enum
+{
+  // XXX: These are the currently used tunables in glibc 2.25 and 2.26.
+  // They are likely to change, so we should be getting them from the debug
+  // symbols instead, via extract-system-config.py.
+  glibc_malloc_arena_max, glibc_malloc_mmap_max, glibc_malloc_mmap_threshold,
+  glibc_malloc_check, glibc_malloc_perturb, glibc_malloc_trim_threshold,
+  glibc_malloc_arena_test, glibc_malloc_top_pad
+} tunable_id_t;
+typedef union
+{
+  // the types of supported tunables
+  int64_t numval;
+  const char *str;
+} tunable_val_t;
+typedef void (*tunable_callback_t) (tunable_val_t *);
+EXPORT void
+__tunable_set_val (tunable_id_t id, void *valp, tunable_callback_t callback)
+{
+  // XXX: Currently the only tunable function I've seen called in practice,
+  // and it just sets them, so an empty stub is fine. Once anything really
+  // makes use of them though, we need to reimplement ~the whole system.
+  return;
+}
 
 static void **
 vdl_dl_error_catch_tsd (void)

--- a/src/external/elf-loader/glibc.c
+++ b/src/external/elf-loader/glibc.c
@@ -219,9 +219,12 @@ glibc_startup_finished (void)
 void
 glibc_initialize (int clktck)
 {
+  // _dl_error_catch_tsd only exists in glibc versions <2.25
+#ifdef CONFIG_DL_ERROR_CATCH_TSD_OFFSET
   void **(*fn) (void) = vdl_dl_error_catch_tsd;
   char *dst = &_rtld_local[CONFIG_DL_ERROR_CATCH_TSD_OFFSET];
   vdl_memcpy ((void *) dst, &fn, sizeof (fn));
+#endif
   char *off = &_rtld_local_ro[CONFIG_RTLD_DL_PAGESIZE_OFFSET];
   int pgsz = system_getpagesize ();
   vdl_memcpy (off, &pgsz, sizeof (pgsz));

--- a/src/external/elf-loader/stage2.c
+++ b/src/external/elf-loader/stage2.c
@@ -328,6 +328,12 @@ stage2_initialize (struct Stage2Input input)
   // We need to do this before relocation because the TLS-type relocations
   // need tls information.
   vdl_tls_file_initialize_main (context->loaded);
+  // We also need to set up the TCB before pthreads relocates
+  unsigned long tcb = vdl_tls_tcb_allocate ();
+  vdl_tls_tcb_initialize (tcb, input.sysinfo);
+  vdl_tls_dtv_allocate (tcb);
+  // configure the current thread to use this TCB as a thread pointer
+  machine_thread_pointer_set (tcb);
 
   // We either setup the GOT for lazy symbol resolution
   // or we perform binding for all symbols now if LD_BIND_NOW is set
@@ -337,12 +343,7 @@ stage2_initialize (struct Stage2Input input)
   // and the dtv. We need to wait post-reloc because the tls
   // template area used to initialize the tls blocks is likely
   // to be modified during relocation processing.
-  unsigned long tcb = vdl_tls_tcb_allocate ();
-  vdl_tls_tcb_initialize (tcb, input.sysinfo);
-  vdl_tls_dtv_allocate (tcb);
   vdl_tls_dtv_initialize (tcb);
-  // configure the current thread to use this TCB as a thread pointer
-  machine_thread_pointer_set (tcb);
 
   // Note that we must invoke this method to notify gdb that we have
   // a valid linkmap only _after_ relocations have been done (if you do

--- a/src/external/elf-loader/test/CMakeLists.txt
+++ b/src/external/elf-loader/test/CMakeLists.txt
@@ -211,5 +211,5 @@ target_link_libraries(test27 -ldl)
 add_test(NAME elfloader-test27 COMMAND /bin/bash ${CMAKE_CURRENT_SOURCE_DIR}/runtest.sh test27 ${CMAKE_CURRENT_SOURCE_DIR})
 
 add_executable(test28 test28.c)
-target_link_libraries(test28 -lpthread -ldl)
+target_link_libraries(test28 vdl -lpthread -ldl)
 add_test(NAME elfloader-test28 COMMAND /bin/bash ${CMAKE_CURRENT_SOURCE_DIR}/runtest.sh test28 ${CMAKE_CURRENT_SOURCE_DIR})

--- a/src/external/elf-loader/vdl-tls.c
+++ b/src/external/elf-loader/vdl-tls.c
@@ -216,33 +216,96 @@ vdl_tls_tcb_initialize (unsigned long tcb, unsigned long sysinfo)
               sizeof (sysinfo));
 }
 
-// This dtv structure needs to be compatible with the one used by the
+// The dtv_t structure needs to be compatible with the one used by the
 // glibc loader. Although it's supposed to be opaque to the glibc or
 // libpthread, it's not. nptl_db reads it to lookup tls variables (it
 // reads dtv[i].value where i >= 1 to find out the address of a target
-// tls block) and libpthread reads dtv[-1].value to find out the size
+// tls block) and libpthread reads dtv[-1] to find out the size
 // of the dtv array and be able to memset it to zeros.
-// dtv[0].value is used as glibc/pthreads' generation counter.
-// The only leeway we have is in the glibc static field which we reuse
-// to store a per-dtvi generation counter.
-// the original in the glibc source can be found in sysdeps/[arch]/nptl/tls.h
-struct dtv_t
+// dtv[0] is used as glibc/pthreads' generation counter.
+// Details depend on the glibc version, see comments below.
+
+struct dtv_meta
 {
-  unsigned long value;
+  unsigned long nptl;
+  //ABI < glibc 2.25
   unsigned long is_static:1;
   unsigned long gen:(sizeof (unsigned long) * 8 - 1);
 };
 
-static inline struct dtv_t *
+struct dtv_ptrs
+{
+  void *value;
+  // ABI >= glibc 2.25
+  void *to_free;
+};
+
+typedef union dtv
+{
+  struct dtv_meta meta;
+  struct dtv_ptrs ptrs;
+} dtv_t;
+
+// some macros to define special values in the dtv
+// macros with "ABI" in them are part of the glibc/pthreads ABI
+#define DTV_ABI_GEN(dtv) dtv[0].meta.gen
+// number of actively used elements
+#define DTV_ABI_SIZE(dtv) dtv[-1].meta.nptl
+// size of the buffer allocated
+#define DTV_MEM_SIZE(dtv) dtv[-1].meta.gen
+// tls for elf-loader to use
+#define DTV_LOCAL_TLS(dtv) dtv[-2].ptrs.value
+
+// we support two ABIs for the dtv, because of the new layout in glibc 2.25
+#if __GLIBC_MINOR__ < 25
+// In older versions, there's a single bit is_static flag we can use,
+// plus store the generation counter in the slack space of the mem-aligned
+// struct. No need for extra structs or functionality.
+typedef dtv_t shadowdtv_t;
+#define DTV_SHADOW_DTV(dtv) dtv
+#define DTV_ALLOCATE_SHADOW(dtv, size)
+#define DTV_MIGRATE_SHADOW(old_dtv, new_dtv, module)
+#define DTV_FREE_SHADOW(dtv)
+#define DTV_ABI_SET_TO_FREE(dtv, module)
+#else
+// From 25 on, the dtv struct replaces the "is_static" flag with a "to_free"
+// field to keep track of the unaligned memory to call free() on.
+// This causes two problems.
+// One: we no longer have room in the dtv to store our metadata.
+// We solve this by adding a "shadow" dtv that stores the fields that used to
+// be in the dtv itself.
+typedef union shadowdtv
+{
+  struct
+  {
+    unsigned long is_static:1; // width isn't for an ABI, it just saves space
+    unsigned long gen:(sizeof (unsigned long) * 8 - 1);
+  } meta;
+} shadowdtv_t;
+#define DTV_SHADOW_DTV(dtv) dtv[-2].ptrs.to_free
+#define DTV_ALLOCATE_SHADOW(dtv, size)          \
+  DTV_SHADOW_DTV(dtv) = vdl_alloc_malloc (size)
+#define DTV_FREE_SHADOW(dtv) vdl_alloc_free (DTV_SHADOW_DTV(dtv))
+#define DTV_MIGRATE_SHADOW(new_dtv, old_dtv, module)    \
+  ((shadowdtv_t *) DTV_SHADOW_DTV(new_dtv))[module] =   \
+    ((shadowdtv_t *) DTV_SHADOW_DTV(old_dtv))[module]
+// Two: glibc's free() isn't our internal free(), and has an incompatible ABI.
+// Since it's us who allocates the TLS, we can't let glibc clean it up.
+// We make sure to always set the "to_free" field to 0, which free() ignores.
+#define DTV_ABI_SET_TO_FREE(dtv, module)        \
+  dtv[module].ptrs.to_free = 0
+#endif
+
+static inline dtv_t *
 get_current_dtv (unsigned long tp)
 {
-  struct dtv_t *dtv;
+  dtv_t *dtv;
   vdl_memcpy (&dtv, (void *) (tp + CONFIG_TCB_DTV_OFFSET), sizeof (dtv));
   return dtv;
 }
 
 static inline void
-set_current_dtv (unsigned long tp, struct dtv_t *dtv)
+set_current_dtv (unsigned long tp, dtv_t *dtv)
 {
   vdl_memcpy ((void *) (tp + CONFIG_TCB_DTV_OFFSET), &dtv, sizeof (dtv));
 }
@@ -251,10 +314,10 @@ void
 vdl_tls_dtv_allocate (unsigned long tcb)
 {
   VDL_LOG_FUNCTION ("tcb=%lu", tcb);
-  struct dtv_t *new_dtv, *current_dtv = get_current_dtv (tcb);
+  dtv_t *new_dtv, *current_dtv = get_current_dtv (tcb);
   // the 3 here is the two entries we put before the dtv, plus a new entry
-  unsigned long needed_size = (3 + g_vdl.tls_n_dtv) * sizeof (struct dtv_t);
-  bool migrate = current_dtv && (current_dtv[-1].gen < needed_size);
+  unsigned long needed_size = (3 + g_vdl.tls_n_dtv) * sizeof (dtv_t);
+  bool migrate = current_dtv && (DTV_MEM_SIZE(current_dtv) < needed_size);
   struct LocalTLS *local_tls;
   if (!current_dtv)
     {
@@ -264,7 +327,7 @@ vdl_tls_dtv_allocate (unsigned long tcb)
     }
   else
     {
-      local_tls = (struct LocalTLS *) current_dtv[-2].value;
+      local_tls = DTV_LOCAL_TLS(current_dtv);
     }
 
   if (!current_dtv || migrate)
@@ -274,8 +337,10 @@ vdl_tls_dtv_allocate (unsigned long tcb)
       // there are 2 entries before the dtv:
       new_dtv++; // our (elf-loader's) thread-local storage
       new_dtv++; // the metadata used by pthreads
-      new_dtv[-2].value = (unsigned long) local_tls;
-      new_dtv[-1].gen = 2 * needed_size;
+      DTV_LOCAL_TLS(new_dtv) = local_tls;
+      DTV_MEM_SIZE(new_dtv) = 2 * needed_size;
+      // Must always be the same size as the real dtv.
+      DTV_ALLOCATE_SHADOW(new_dtv, 2 * needed_size);
       set_current_dtv (tcb, new_dtv);
     }
   else
@@ -283,17 +348,18 @@ vdl_tls_dtv_allocate (unsigned long tcb)
       new_dtv = current_dtv;
     }
 
-  new_dtv[-1].value = g_vdl.tls_n_dtv;
-  new_dtv[0].value = 0;
-  new_dtv[0].gen = g_vdl.tls_gen;
+  DTV_ABI_SIZE(new_dtv) = g_vdl.tls_n_dtv;
+  new_dtv[0].meta.nptl = 0;
+  DTV_ABI_GEN(new_dtv) = g_vdl.tls_gen;
 
   if (migrate)
     {
       unsigned long module;
       // copy over the data from the old dtv into the new one.
-      for (module = 1; module <= current_dtv[-1].value; module++)
+      for (module = 1; module <= DTV_ABI_SIZE(current_dtv); module++)
         {
           new_dtv[module] = current_dtv[module];
+          DTV_MIGRATE_SHADOW(new_dtv, current_dtv, module);
         }
       // clear the old dtv
       vdl_alloc_free (&current_dtv[-2]);
@@ -304,8 +370,8 @@ void
 vdl_tls_dtv_initialize (unsigned long tcb)
 {
   VDL_LOG_FUNCTION ("tcb=%lu", tcb);
-  struct dtv_t *dtv;
-  vdl_memcpy (&dtv, (void *) (tcb + CONFIG_TCB_DTV_OFFSET), sizeof (dtv));
+  dtv_t *dtv = get_current_dtv (tcb);
+  shadowdtv_t *shadow_dtv = DTV_SHADOW_DTV(dtv);
   // allocate a dtv for the set of tls blocks needed now
 
   struct VdlFile *cur;
@@ -316,9 +382,9 @@ vdl_tls_dtv_initialize (unsigned long tcb)
           // setup the dtv to point to the tls block
           if (cur->tls_is_static)
             {
-              signed long dtvi = tcb + cur->tls_offset;
-              dtv[cur->tls_index].value = dtvi;
-              dtv[cur->tls_index].is_static = 1;
+              void *dtvi = (void *)tcb + cur->tls_offset;
+              dtv[cur->tls_index].ptrs.value = dtvi;
+              shadow_dtv[cur->tls_index].meta.is_static = 1;
               // copy the template in the module tls block
               vdl_memcpy ((void *) dtvi, (void *) cur->tls_tmpl_start,
                           cur->tls_tmpl_size);
@@ -327,14 +393,16 @@ vdl_tls_dtv_initialize (unsigned long tcb)
             }
           else
             {
-              dtv[cur->tls_index].value = 0;    // unallocated
-              dtv[cur->tls_index].is_static = 0;
+              dtv[cur->tls_index].ptrs.value = 0;    // unallocated
+              shadow_dtv[cur->tls_index].meta.is_static = 0;
             }
-          dtv[cur->tls_index].gen = cur->tls_tmpl_gen;
+          DTV_ABI_SET_TO_FREE(dtv, cur->tls_index);
+          shadowdtv_t *shadow_dtv = DTV_SHADOW_DTV(dtv);
+          shadow_dtv[cur->tls_index].meta.gen = cur->tls_tmpl_gen;
         }
     }
   // initialize its generation counter
-  dtv[0].gen = g_vdl.tls_gen;
+  DTV_ABI_GEN(dtv) = g_vdl.tls_gen;
 }
 
 inline struct LocalTLS *
@@ -343,10 +411,10 @@ vdl_tls_get_local_tls (void)
   if (g_vdl.tp_set)
     {
       unsigned long tp = machine_thread_pointer_get ();
-      struct dtv_t *dtv = get_current_dtv (tp);
+      dtv_t *dtv = get_current_dtv (tp);
       if (dtv)
         {
-          return (struct LocalTLS *) dtv[-2].value;
+          return (struct LocalTLS *) DTV_LOCAL_TLS(dtv);
         }
     }
   return 0;
@@ -373,31 +441,33 @@ void
 vdl_tls_dtv_deallocate (unsigned long tcb)
 {
   VDL_LOG_FUNCTION ("tcb=%lu", tcb);
-  struct dtv_t *dtv = get_current_dtv (tcb);
+  dtv_t *dtv = get_current_dtv (tcb);
+  shadowdtv_t *shadow_dtv = DTV_SHADOW_DTV(dtv);
 
-  unsigned long dtv_size = dtv[-1].value;
+  unsigned long dtv_size = DTV_ABI_SIZE(dtv);
   unsigned long module;
   for (module = 1; module <= dtv_size; module++)
     {
-      if (dtv[module].value == 0)
+      if (dtv[module].ptrs.value == 0)
         {
           // this was an unallocated entry
           continue;
         }
-      if (dtv[module].is_static)
+      if (shadow_dtv[module].meta.is_static)
         {
           // this was a static entry so, we don't
           // have anything to free here.
           continue;
         }
       // this was not a static entry
-      unsigned long *dtvi = (unsigned long *) dtv[module].value;
+      unsigned long *dtvi = (unsigned long *) dtv[module].ptrs.value;
       vdl_alloc_free (&dtvi[-1]);
     }
   // If we could, we would free the allocator associated with this thread now.
   // But it's possible that it has allocated memory that will be used/freed
   // later, on some other thread, so we can't.
-  vdl_alloc_free ((struct LocalTLS *) dtv[-2].value);
+  vdl_alloc_free ((struct LocalTLS *) DTV_LOCAL_TLS(dtv));
+  DTV_FREE_SHADOW(dtv);
   vdl_alloc_free (&dtv[-2]);
 }
 
@@ -414,7 +484,7 @@ vdl_tls_tcb_deallocate (unsigned long tcb)
 // For now, we'll just remove it until we actually need to implement it.
 // When we do implement it, move this logic into the deinitialize functions.
 static inline void
-vdl_tls_dtv_update_current (__attribute__ ((unused)) struct dtv_t *dtv,
+vdl_tls_dtv_update_current (__attribute__ ((unused)) dtv_t *dtv,
                             __attribute__ ((unused)) unsigned long dtv_size)
 {
   /*
@@ -456,15 +526,17 @@ vdl_tls_dtv_update_current (__attribute__ ((unused)) struct dtv_t *dtv,
 }
 
 static inline void
-vdl_tls_dtv_update_new (struct dtv_t *new_dtv, unsigned long dtv_size,
+vdl_tls_dtv_update_new (dtv_t *new_dtv, unsigned long dtv_size,
                         unsigned long new_dtv_size, signed long tcb)
 {
   unsigned long module;
+  shadowdtv_t *new_shadow_dtv = DTV_SHADOW_DTV(new_dtv);
   for (module = dtv_size + 1; module <= new_dtv_size; module++)
     {
-      new_dtv[module].value = 0;
-      new_dtv[module].gen = 0;
-      new_dtv[module].is_static = 0;
+      new_dtv[module].ptrs.value = 0;
+      DTV_ABI_SET_TO_FREE(new_dtv, module);
+      new_shadow_dtv[module].meta.gen = 0;
+      new_shadow_dtv[module].meta.is_static = 0;
       struct VdlFile *file = find_file_by_module (module);
       if (file == 0)
         {
@@ -475,26 +547,26 @@ vdl_tls_dtv_update_new (struct dtv_t *new_dtv, unsigned long dtv_size,
         }
       if (file->tls_is_static)
         {
-          signed long dtvi = tcb + file->tls_offset;
-          new_dtv[file->tls_index].value = dtvi;
-          new_dtv[file->tls_index].is_static = 1;
-          new_dtv[file->tls_index].gen = file->tls_tmpl_gen;
+          void *dtvi = (void *)tcb + file->tls_offset;
+          new_dtv[file->tls_index].ptrs.value = dtvi;
+          new_shadow_dtv[file->tls_index].meta.is_static = 1;
+          new_shadow_dtv[file->tls_index].meta.gen = file->tls_tmpl_gen;
           // copy the template in the module tls block
-          vdl_memcpy ((void *) dtvi, (void *) file->tls_tmpl_start,
+          vdl_memcpy (dtvi, (void *) file->tls_tmpl_start,
                       file->tls_tmpl_size);
-          vdl_memset ((void *) (dtvi + file->tls_tmpl_size), 0,
+          vdl_memset (dtvi + file->tls_tmpl_size, 0,
                       file->tls_init_zero_size);
         }
     }
 }
 
 static void
-vdl_tls_dtv_update_given (unsigned long tp, struct dtv_t *dtv)
+vdl_tls_dtv_update_given (unsigned long tp, dtv_t *dtv)
 {
   VDL_LOG_FUNCTION ("");
-  unsigned long dtv_size = dtv[-1].value;
+  unsigned long dtv_size = DTV_ABI_SIZE(dtv);
 
-  if (dtv[0].gen == g_vdl.tls_gen)
+  if (DTV_ABI_GEN(dtv) == g_vdl.tls_gen)
     {
       return;
     }
@@ -507,19 +579,19 @@ vdl_tls_dtv_update_given (unsigned long tp, struct dtv_t *dtv)
     {
       // we have a big-enough dtv so, now that it's uptodate,
       // update the generation
-      dtv[0].gen = g_vdl.tls_gen;
+      DTV_ABI_GEN(dtv) = g_vdl.tls_gen;
       return;
     }
 
   // the size of the new dtv is bigger than the
   // current dtv. We need a newly-sized dtv
   vdl_tls_dtv_allocate (tp);
-  struct dtv_t *new_dtv = get_current_dtv (tp);
-  unsigned long new_dtv_size = new_dtv[-1].value;
+  dtv_t *new_dtv = get_current_dtv (tp);
+  unsigned long new_dtv_size = DTV_ABI_SIZE(new_dtv);
   // then, initialize the new area in the new dtv
   vdl_tls_dtv_update_new (new_dtv, dtv_size, new_dtv_size, tp);
   // now that the dtv is updated, update the generation
-  new_dtv[0].gen = g_vdl.tls_gen;
+  DTV_ABI_GEN(new_dtv) = g_vdl.tls_gen;
 }
 
 void
@@ -527,7 +599,7 @@ vdl_tls_dtv_update (void)
 {
   unsigned long tp = machine_thread_pointer_get ();
   read_lock (g_vdl.tls_lock);
-  struct dtv_t *dtv = get_current_dtv (tp);
+  dtv_t *dtv = get_current_dtv (tp);
   vdl_tls_dtv_update_given (tp, dtv);
   read_unlock (g_vdl.tls_lock);
 }
@@ -536,12 +608,12 @@ unsigned long
 vdl_tls_get_addr_fast (unsigned long module, unsigned long offset)
 {
   unsigned long tp = machine_thread_pointer_get ();
-  struct dtv_t *dtv = get_current_dtv (tp);
-  if (dtv[0].gen == g_vdl.tls_gen && dtv[module].value != 0)
+  dtv_t *dtv = get_current_dtv (tp);
+  if (DTV_ABI_GEN(dtv) == g_vdl.tls_gen && dtv[module].ptrs.value != 0)
     {
       // our dtv is really uptodate _and_ the requested module block
       // has been already initialized.
-      return dtv[module].value + offset;
+      return dtv[module].ptrs.value + offset;
     }
   // either we need to update the dtv or we need to initialize
   // the dtv entry to point to the requested module block
@@ -560,8 +632,9 @@ vdl_tls_get_addr_slow (unsigned long module, unsigned long offset)
       return addr;
     }
   unsigned long tp = machine_thread_pointer_get ();
-  struct dtv_t *dtv = get_current_dtv (tp);
-  if (dtv[0].gen == g_vdl.tls_gen && dtv[module].value == 0)
+  dtv_t *dtv = get_current_dtv (tp);
+  shadowdtv_t *shadow_dtv = DTV_SHADOW_DTV(dtv);
+  if (DTV_ABI_GEN(dtv) == g_vdl.tls_gen && dtv[module].ptrs.value == 0)
     {
       // the dtv is uptodate but the requested module block
       // has not been initialized already
@@ -578,12 +651,13 @@ vdl_tls_get_addr_slow (unsigned long module, unsigned long offset)
       vdl_memset ((void *) (((unsigned long) dtvi) + file->tls_tmpl_size),
                   0, file->tls_init_zero_size);
       // finally, update the dtv
-      dtv[module].value = (unsigned long) dtvi;
-      dtv[module].gen = file->tls_tmpl_gen;
-      dtv[module].is_static = 0;
+      dtv[module].ptrs.value = dtvi;
+      DTV_ABI_SET_TO_FREE(dtv, module);
+      shadow_dtv[module].meta.gen = file->tls_tmpl_gen;
+      shadow_dtv[module].meta.is_static = 0;
       // and return the requested value
       read_unlock (g_vdl.tls_lock);
-      return dtv[module].value + offset;
+      return dtv[module].ptrs.value + offset;
     }
   // we know for sure that the dtv is _not_ uptodate now
   vdl_tls_dtv_update_given (tp, dtv);
@@ -598,8 +672,8 @@ struct SwapArgs
 {
   unsigned long t1;
   unsigned long t2;
-  struct dtv_t *dtv1;
-  struct dtv_t *dtv2;
+  dtv_t *dtv1;
+  dtv_t *dtv2;
 };
 
 void *
@@ -629,9 +703,11 @@ vdl_tls_swap_file (void *data, void *aux)
   // make sure we're not trying to swap the gen counter
   if ((module = file->tls_index) > 0)
     {
-      struct dtv_t tmp_dtv = args->dtv1[module];
+      dtv_t tmp_dtv = args->dtv1[module];
       args->dtv1[module] = args->dtv2[module];
       args->dtv2[module] = tmp_dtv;
+      // we don't need to swap the shadow dtvs because we should only be
+      // swapping after an update, so metadata it stores should be the same
     }
   return 0;
 }
@@ -640,8 +716,8 @@ void
 vdl_tls_swap_context (struct VdlContext *context, unsigned long t1, unsigned long t2)
 {
   write_lock (g_vdl.tls_lock);
-  struct dtv_t *dtv1 = get_current_dtv (t1);
-  struct dtv_t *dtv2 = get_current_dtv (t2);
+  dtv_t *dtv1 = get_current_dtv (t1);
+  dtv_t *dtv2 = get_current_dtv (t2);
   // make sure we're not copying from/to uninitialized/unallocated memory
   vdl_tls_dtv_update_given (t1, dtv1);
   vdl_tls_dtv_update_given (t2, dtv2);

--- a/src/external/elf-loader/x86_64/stage0.S
+++ b/src/external/elf-loader/x86_64/stage0.S
@@ -22,7 +22,14 @@ L1:	lea -L1(%rdi),%rdi
 	push %rdi
 	# pass pointer to TrampolineInformation to stage1
 	lea 8(%rsp),%rdi
+        # save the stack pointer for when we come back
+        # (r12 is preserved by functions and syscalls)
+        movq %rsp, %r12
+        # align the stack to 16 bytes
+        andq $-16, %rsp
 	call stage1
+        # get the stack pointer back
+        movq %r12, %rsp
 	# read dl_fini field of TrampolineInformation
 	mov 32(%rsp),%rdx
 	# read entry_point field of TrampolineInformation
@@ -35,4 +42,3 @@ L1:	lea -L1(%rdi),%rdi
 	# trick from glibc: crash if we return here. Should not
 	# happen ever
 	hlt
-	

--- a/src/test/tcp/shd-test-tcp.c
+++ b/src/test/tcp/shd-test-tcp.c
@@ -19,6 +19,7 @@
 #include <sys/epoll.h>
 #include <sys/select.h>
 #include <fcntl.h>
+#include <sys/uio.h>
 
 #define USAGE "USAGE: 'shd-test-tcp iomode type'; iomode=('blocking'|'nonblocking-poll'|'nonblocking-epoll'|'nonblocking-select') type=('client' server_ip|'server')"
 #define MYLOG(...) _mylog(__FILE__, __LINE__, __FUNCTION__, __VA_ARGS__)

